larana/OpticalDetector/SimPhotonCounter_module.cc

```diff
@@ -480,107 +491,96 @@ namespace opdet {
+            }
 
-          // Fill per event tree
-          if(fMakeOpDetEventsTree) fTheEventTree->Fill();
+            // Fill per event tree
+            if(fMakeOpDetEventsTree) fTheEventTree->Fill();
 
-          // Give per event output
-          if(fVerbosity >1) std::cout<<"OpDetResponseInterface PerEvent : Event "<<fEventID<<" All " << fCountOpDetAll << " Det " <<fCountOpDetDetected<<std::endl;
+            // Give per event output
+            if(fVerbosity >1) std::cout<<"OpDetResponseInterface PerEvent : Event "<<fEventID<<" All " << fCountOpDetAll << " Det " <<fCountOpDetDetected<<std::endl;
 
-        }
-        else
-        {
-          // if empty OpDet hit collection,
-          // add an empty record to the per event tree
-          if(fMakeOpDetEventsTree) fTheEventTree->Fill();
-        }
-        if(fMakeLightAnalysisTree) {
-          std::cout<<"Filling the analysis tree"<<std::endl;
-          //---------------Filling the analysis tree-----------:
-          fRun = evt.run();
-          std::vector<double> this_xyz;
-
-          //loop over the particles
-          for(size_t i_p=0; i_p < mcpartVec.size(); i_p++){
-
-	    const simb::MCParticle pPart = mcpartVec[i_p];
-	    if(pPart.Process() == "primary")
-	      fEnergy = pPart.E();
-
-            //resetting the vectors
-            fstepPositions.clear();
-            fstepTimes.clear();
-            fSignalsvuv.clear();
-            fSignalsvis.clear();
-            fdEdx = -1.;
-            //filling the tree fields
-	    fTrackID = pPart.TrackId();
-	    fpdg = pPart.PdgCode();
-	    fmotherTrackID = pPart.Mother();
-	    fdEdx = totalEnergy_track[fTrackID];
-	    fSignalsvuv = fSignals_vuv[fTrackID];
-	    fSignalsvis = fSignals_vis[fTrackID];
-	    fProcess = pPart.Process();
-            //filling the center positions of each step
-            for(size_t i_s=1; i_s < pPart.NumberTrajectoryPoints(); i_s++){
-	      this_xyz.clear();
-	      this_xyz.resize(3);
-	      this_xyz[0] = pPart.Position(i_s).X();
-	      this_xyz[1] = pPart.Position(i_s).Y();
-	      this_xyz[2] = pPart.Position(i_s).Z();
-	      fstepPositions.push_back(this_xyz);
-	      fstepTimes.push_back(pPart.Position(i_s).T());
-	    }
-            //filling the tree per track
-            fLightAnalysisTree->Fill();
           }
-	}
+          else
+          {
+            // if empty OpDet hit collection,
+            // add an empty record to the per event tree
+            if(fMakeOpDetEventsTree) fTheEventTree->Fill();
+          }
+          if(fMakeLightAnalysisTree) {
+            std::cout<<"Filling the analysis tree"<<std::endl;
+            //---------------Filling the analysis tree-----------:
+            fRun = evt.run();
+            std::vector<double> this_xyz;
+
+            //loop over the particles
+            for(simb::MCParticle const& pPart: *mcpartVec){
```

Comment:
Are you certain that `mcpartVec` is always non-null in this context?

Reply:
I think we agree this code follows the anamorphous-growing design style.
The summary is:
* we have feature A, B... optional and regulated by `flagA`, `flagB`, etc.
* the code performs some initialisation of `stuffA`, and `stuffB` if `flagB` is set, ...
* the code enters a loop; on each iteration, it does A if `flagA` is set, it does B if `flagB` is set, ...

In fact, what I added is to skip `stuffA` if `flagA` is not set. I am not attempting to redesign the module: anything short of redesigning `PhotonLibrary`, `PhotonVisibilityService` and `LightSource` module is going to be an unsatisfactory compromise, and the amount of work for this redesign is weeks.

So here the feature I am playing with is ultimately controlled by `fMakeLightAnalysisTree` (the new branding of `flagA`). That controls both the resources in input (`mcpartVec`) and in output (`fLightAnalysisTree`). As of now, I am decently confident that the interaction among these parts is correct.

In the new commit I am proposing a slightly less fragmented initialisation and the explicit control of all the code branches pertaining this feature by `fMakeLightAnalysisTree`. I also document the requirement of the two controlled structures via assertions.

Is this an acceptable compromise?

Commit 0bdd959.

Reply:
That's better, Gianluca.  Thanks.